substrate/frame/contracts/src/exec.rs

```diff
@@ -907,13 +907,19 @@ where
 			// Every non delegate call or instantiate also optionally transfers the balance.
 			self.initial_transfer()?;
 
-			let call_span =
-				T::Debug::new_call_span(executable.code_hash(), entry_point, &input_data);
-
-			// Call into the Wasm blob.
-			let output = executable
-				.execute(self, &entry_point, input_data)
-				.map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?;
+			let contract_address = &top_frame!(self).account_id;
+
+			let call_span = T::Debug::new_call_span(contract_address, entry_point, &input_data);
+			let interception_result =
+				T::Debug::intercept_call(contract_address, entry_point, &input_data);
+			let output = match interception_result {
+				// Call into the Wasm blob.
+				CallInterceptorResult::Proceed => executable
+					.execute(self, &entry_point, input_data)
+					.map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?,
+				// Return the output of the interceptor.
+				CallInterceptorResult::Stop(output) => output?,
+			};
```

Comment:
Maybe we can move the details of intercept_call to the trait implementation, so we don't leak the struct that could then be internal to the implementation. 

We could move that behind a feature-flag like this 
```rust
let output = if cfg!(feature = "unsafe-debug") {
    T::Debug::wrap_execute(self, contract_address, &entry_point, input_data)
} else {
    executable.execute(self, &entry_point, input_data)
}.map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?
```

or maybe simply

```rust
let output = T::Debug::wrap_execute(self, contract_address, &entry_point, input_data)
  .map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?
```

Reply:
> We could move that behind a feature-flag like this

we have removed the feature flag, haven't we?

> we don't leak the struct

what if we return just `Result<(), ExecResult>` instead of the current new enum type? same semantics, no new types

Reply:
> what if we return just `Result<(), ExecResult>` instead of the current new enum type? same semantics, no new types

Or `Option<ExecResult>` which is less confusing to the implementor of the trait but would require a bit more work here to handle it properly.

Reply:
> we have removed the feature flag, haven't we?

We can always re-introduce it. For tracing we needed to have basic traces regardless of the environment, so we could get away without introducing a feature flag. Intercepting calls seems more like a dev only feature, so I think it could make sense to have that gated behind a feature flag.

Moving it all behind a ` T::Debug::wrap_execute` that would just call the actual code with the default `()` Debug trait might be good alternative as well. @athei any thought on this?

> what if we return just Result<(), ExecResult> instead of the current new enum type? same semantics, no new types

I think that's what I am suggesting with the code snippet above.

Reply:
> I think that's what I am suggesting with the code snippet above.

That's not how we saw it. You suggested (assuming it's the 2nd snippet):
```rust
let output = T::Debug::wrap_execute(self, contract_address, &entry_point, input_data)
  .map_err(|e| ExecError { error: e.error, origin: ErrorOrigin::Callee })?
```
which looked like the `wrap_execute` should also call `execute`? 

I think we can leave the `T::Debug::intercept_call` as-is but change the return type.

if we agree to the general idea I think we can pin down the details (method name and return type).

Reply:
> which looked like the wrap_execute should also call execute?
Yes right

> I think we can leave the T::Debug::intercept_call as-is but change the return type.
So you would do something like
```rust
if let Some(result) = T::Debug::intercept_call(contract_address, entry_point, &input_data) {
  result
} else {
// current logic
}
```

Yes I guess we can start with this

Reply:
so all in all I've removed the new `CallInterceptorResult` enum in favor of `Option`... does it meet the consensus above?